java-frontend/src/main/java/org/sonar/java/model/statement/TryStatementTreeImpl.java

```diff
@@ -81,13 +85,33 @@ public TryStatementTreeImpl(
     this.tryToken = tryToken;
     this.openParenToken = openParenToken;
     this.resources = resources;
+    this.variableTreeResources = filterVariableTreeResources(resources);
     this.closeParenToken = closeParenToken;
     this.block = block;
     this.catches = getCatches(catches);
     this.finallyKeyword = null;
     this.finallyBlock = null;
   }
 
+  private static ListTree filterVariableTreeResources(ResourceListTreeImpl resources) {
+    ImmutableList.Builder<Tree> filteredResources = ImmutableList.builder();
+    ImmutableList.Builder<SyntaxToken> filteredSeparators = ImmutableList.builder();
+    Iterator<SyntaxToken> separators = resources.separators().iterator();
+    for (Tree resource : resources) {
+      SyntaxToken separator = null;
+      if (separators.hasNext()) {
+        separator = separators.next();
+      }
+      if (resource.is(Kind.VARIABLE)) {
+        filteredResources.add(resource);
+        if (separator != null) {
+          filteredSeparators.add(separator);
+        }
+      }
+    }
+    return new ResourceListTreeImpl(filteredResources.build(), filteredSeparators.build());
```

Comment:
I think a more elegant approach would be to directly modify the `ResourceListTreeImpl` in order to keep a distinct list of `VariableTree` on its side, which would be created at instantiation time as well. That way, we would keep that logic in one place and won't have to use a ListTree with unspecified type arguments.